atomic_reactor/operator_util.py

```diff
@@ -0,0 +1,238 @@
+
+
+class OperatorCSV(object):
+    """
+    A single ClusterServiceVersion file in an operator manifest.
+
+    Can find and replace pullspecs for container images in predefined locations.
+    """
+
+    def __init__(self, path, data):
+        """
+        Initialize an OperatorCSV
+
+        :param path: Path where data was found or where it should be written
+        :param data: ClusterServiceVersion yaml data
+        """
+        if data.get("kind") != OPERATOR_CSV_KIND:
+            raise NotOperatorCSV("Not a ClusterServiceVersion")
+        self.path = path
+        self.data = data
+
+    @classmethod
+    def from_file(cls, path):
+        """
+        Make an OperatorCSV from a file
+
+        :param path: Path to file
+        :return: OperatorCSV
+        """
+        with open(path) as f:
+            data = yaml.load(f)
+            return cls(path, data)
+
+    def dump(self):
+        """
+        Write data to file (preserves comments)
+        """
+        with open(self.path, "w") as f:
+            yaml.dump(self.data, f)
+
+    def get_pullspecs(self):
+        """
+        Find pullspecs for containers, related images and RELATED_IMAGE_*
+        environment variables in containers.
+
+        :return: set of ImageName pullspecs
+        """
+        pullspecs = self._get_related_image_pullspecs()
+        pullspecs.update(self._get_env_pullspecs())
+        pullspecs.update(self._get_container_image_pullspecs())
+        return pullspecs
+
+    def replace_pullspecs(self, replacement_pullspecs):
+        """
+        Replace pullspecs for containers, related images and RELATED_IMAGE_*
+        environment variables in containers.
+
+        :param replacement_pullspecs: mapping of pullspec -> replacement
+        """
+        self._replace_related_image_pullspecs(replacement_pullspecs)
+        self._replace_env_pullspecs(replacement_pullspecs)
+        self._replace_container_image_pullspecs(replacement_pullspecs)
+
+    def replace_pullspecs_everywhere(self, replacement_pullspecs):
+        """
+        Replace all pullspecs found anywhere in data
+
+        :param replacement_pullspecs: mapping of pullspec -> replacment
+        """
+        log_template = "%(path)s - Replaced pullspec: %(old)s -> %(new)s"
+        for k in self.data:
+            self._replace_pullspecs_everywhere(self.data, k, replacement_pullspecs, log_template)
+
+    def _get_related_images(self):
+        related_images_path = ("spec", "relatedImages")
+        return chain_get(self.data, related_images_path, default=[])
+
+    def _get_containers(self):
+        deployments_path = ("spec", "install", "spec", "deployments")
+        deployments = chain_get(self.data, deployments_path, default=[])
+
+        containers_path = ("spec", "template", "spec", "containers")
+        containers = [
+            c for d in deployments for c in chain_get(d, containers_path, default=[])
+        ]
+        return containers
+
+    def _get_envs(self):
```

Comment:
maybe `get_related_image_envs` ?

Reply:
Renamed, also realized test for "valueFrom" in env definition was missing, added